File: demo/serve.mjs
// Minimal static file server that sets the Cross-Origin-Opener-Policy and
// Cross-Origin-Embedder-Policy headers ffmpeg.wasm needs for
// SharedArrayBuffer. Plain `python -m http.server` or `file://` won't set
// these — that's the #1 reason a fresh ffmpeg.wasm setup silently fails
// with "SharedArrayBuffer is not defined".
//
// Usage: node demo/serve.mjs  (serves the project root on http://localhost:8080)

import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const port = process.env.PORT ? Number(process.env.PORT) : 8080;

const mimeTypes = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript",
  ".mjs": "text/javascript",
  ".json": "application/json",
  ".wasm": "application/wasm",
  ".css": "text/css",
  ".map": "application/json",
  ".png": "image/png",
  ".wav": "audio/wav",
  ".mp4": "video/mp4",
};

const server = http.createServer((req, res) => {
  const parsedUrl = new URL(req.url ?? "/", `http://${req.headers.host || "127.0.0.1"}`);
  const requestPath = decodeURIComponent(parsedUrl.pathname);

  // Proxy route for images
  if (requestPath === "/api/proxy-image") {
    const targetUrl = parsedUrl.searchParams.get("url");
    if (!targetUrl) {
      res.writeHead(400, { "Content-Type": "text/plain" });
      res.end("Missing url param");
      return;
    }

    fetch(targetUrl, {
      headers: {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
      }
    })
      .then(async (resp) => {
        const buffer = await resp.arrayBuffer();
        res.setHeader("Cross-Origin-Opener-Policy", "same-origin");
        res.setHeader("Cross-Origin-Embedder-Policy", "credentialless");
        res.setHeader("Cross-Origin-Resource-Policy", "cross-origin");
        res.setHeader("Access-Control-Allow-Origin", "*");
        res.setHeader("Content-Type", resp.headers.get("content-type") || "image/jpeg");
        res.writeHead(resp.status);
        res.end(Buffer.from(buffer));
      })
      .catch((err) => {
        res.writeHead(502, { "Content-Type": "text/plain" });
        res.end(`Proxy error: ${err.message}`);
      });
    return;
  }

  let filePath = path.join(root, requestPath);
  if (filePath.endsWith(path.sep)) filePath = path.join(filePath, "index.html");

  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404, { "Content-Type": "text/plain" });
      res.end(`Not found: ${req.url}`);
      return;
    }

    const ext = path.extname(filePath);
    res.setHeader("Cross-Origin-Opener-Policy", "same-origin");
    res.setHeader("Cross-Origin-Embedder-Policy", "credentialless");
    res.setHeader("Cross-Origin-Resource-Policy", "cross-origin");
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Content-Type", mimeTypes[ext] ?? "application/octet-stream");
    res.writeHead(200);
    res.end(data);
  });
});

server.listen(port, () => {
  console.log(`Serving ${root} at http://localhost:${port} (COOP/COEP enabled)`);
  console.log(`Open http://localhost:${port}/demo/index.html`);
});

File: demo/serve.py
"""
Smart Video Engine - Local Dev Server with COOP/COEP Headers & Image Proxy
========================================================================
"""

import http.server
import socketserver
import os
import time
import urllib.parse
import urllib.request
import sys

PORT = int(os.environ.get("PORT", 8080))
ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))

_no_proxy_opener = urllib.request.build_opener(urllib.request.ProxyHandler({}))

MIME_TYPES = {
    ".html": "text/html; charset=utf-8",
    ".js":   "text/javascript; charset=utf-8",
    ".mjs":  "text/javascript; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".wasm": "application/wasm",
    ".css":  "text/css; charset=utf-8",
    ".map":  "application/json; charset=utf-8",
    ".png":  "image/png",
    ".jpg":  "image/jpeg",
    ".jpeg": "image/jpeg",
    ".webp": "image/webp",
    ".wav":  "audio/wav",
    ".mp4":  "video/mp4",
    ".svg":  "image/svg+xml",
}


class SmartVideoServerHandler(http.server.SimpleHTTPRequestHandler):
    extensions_map = http.server.SimpleHTTPRequestHandler.extensions_map.copy()
    extensions_map.update({
        ".html": "text/html; charset=utf-8",
        ".js":   "text/javascript; charset=utf-8",
        ".mjs":  "text/javascript; charset=utf-8",
        ".json": "application/json; charset=utf-8",
        ".wasm": "application/wasm",
        ".css":  "text/css; charset=utf-8",
        ".map":  "application/json; charset=utf-8",
        ".png":  "image/png",
        ".jpg":  "image/jpeg",
        ".jpeg": "image/jpeg",
        ".webp": "image/webp",
        ".wav":  "audio/wav",
        ".mp4":  "video/mp4",
        ".svg":  "image/svg+xml",
        "":      "application/octet-stream",
    })

    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "credentialless")
        self.send_header("Cross-Origin-Resource-Policy", "cross-origin")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, HEAD, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "*")
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(200)
        self.end_headers()

    def do_GET(self):
        parsed = urllib.parse.urlparse(self.path)

        if parsed.path == "/api/proxy-image":
            query = urllib.parse.parse_qs(parsed.query)
            target_url = query.get("url", [None])[0]
            if not target_url:
                self.send_error(400, "Missing 'url' parameter")
                return

            # Retry up to 3 times for robustness
            for attempt in range(3):
                try:
                    req = urllib.request.Request(target_url, headers={
                        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
                        "Accept": "image/avif,image/webp,image/apng,image/*,*/*;q=0.8",
                    })
                    with _no_proxy_opener.open(req, timeout=90) as resp:
                        data = resp.read()
                        content_type = resp.headers.get("Content-Type", "image/jpeg")

                        if len(data) > 0:
                            self.send_response(200)
                            self.send_header("Content-Type", content_type)
                            self.send_header("Content-Length", str(len(data)))
                            self.send_header("Cache-Control", "public, max-age=86400")
                            self.end_headers()
                            self.wfile.write(data)
                            return
                except Exception as e:
                    print(f"  [proxy] Attempt {attempt+1} failed: {e}")
                    time.sleep(1)

            # If all retries fail, return a 1x1 dark pixel or error
            self.send_error(502, "Failed to retrieve image after retries")
            return

        if parsed.path in ["", "/", "/index.html"]:
            self.send_response(302)
            self.send_header("Location", "/demo/index.html")
            self.end_headers()
            return

        super().do_GET()

    def guess_type(self, path):
        ext = os.path.splitext(path)[1].lower()
        return MIME_TYPES.get(ext, "application/octet-stream")

    def log_message(self, format, *args):
        pass


class ThreadingServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    allow_reuse_address = True
    daemon_threads = True


def run():
    with ThreadingServer(("", PORT), SmartVideoServerHandler) as httpd:
        print(f"\n========================================================")
        print(f"  Smart Video Engine Demo Server (threaded)")
        print(f"  Root:    {ROOT}")
        print(f"  URL:     http://localhost:{PORT}/demo/index.html")
        print(f"========================================================\n")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nServer stopped.")


if __name__ == "__main__":
    run()
